ash-swapchain/src/lib.rs

```diff
@@ -0,0 +1,415 @@
+            if !self.needs_rebuild {
+                match fp
+                    .swapchain
+                    .acquire_next_image(self.handle, !0, acquire, vk::Fence::null())
+                {
+                    Ok((index, suboptimal)) => {
+                        self.needs_rebuild = suboptimal;
+                        let invalidate_images =
+                            self.frames[frame_index].generation != self.generation;
+                        self.frames[frame_index].generation = self.generation;
+                        self.frame_index = next_frame_index;
+                        fp.device
+                            .reset_fences(&[self.frames[frame_index].complete])
+                            .unwrap();
+                        return Ok(AcquiredFrame {
+                            image_index: index as usize,
+                            frame_index,
+                            ready: acquire,
+                            complete: self.frames[frame_index].complete,
+                            invalidate_images,
+                        });
+                    }
+                    Err(vk::Result::ERROR_OUT_OF_DATE_KHR) => {}
+                    Err(e) => return Err(e),
+                }
+            };
+            self.needs_rebuild = true;
+
+            // Rebuild swapchain
+            let surface_capabilities = fp
+                .surface
+                .get_physical_device_surface_capabilities(self.physical_device, self.surface)?;
+            self.extent = match surface_capabilities.current_extent.width {
+                // If Vulkan doesn't know, the windowing system probably does. Known to apply at
+                // least to Wayland.
+                std::u32::MAX => vk::Extent2D {
+                    width: self.extent.width,
+                    height: self.extent.height,
+                },
+                _ => surface_capabilities.current_extent,
+            };
+            let pre_transform = if surface_capabilities
+                .supported_transforms
+                .contains(vk::SurfaceTransformFlagsKHR::IDENTITY)
+            {
+                vk::SurfaceTransformFlagsKHR::IDENTITY
+            } else {
+                surface_capabilities.current_transform
+            };
+            let present_mode = fp
+                .surface
+                .get_physical_device_surface_present_modes(self.physical_device, self.surface)?
+                .iter()
+                .filter_map(|&mode| {
+                    Some((
+                        mode,
+                        self.options
+                            .present_mode_preference
+                            .iter()
+                            .position(|&pref| pref == mode)?,
+                    ))
+                })
+                .min_by_key(|&(_, priority)| priority)
+                .map(|(mode, _)| mode)
+                .ok_or(vk::Result::ERROR_OUT_OF_DATE_KHR)?;
+
+            let desired_image_count =
```

Comment:
Why `min_image_count + 1`?

Reply:
This is standard practice to reduce sync coupling with the WSI. See also e.g. https://github.com/charles-lunarg/vk-bootstrap/blob/master/src/VkBootstrap.cpp#L1763.